java/instrumentation/openinference-instrumentation-langchain4j/src/main/java/io/openinference/instrumentation/langchain4j/LangChain4jModelListener.java

```diff
@@ -0,0 +1,295 @@
+            for (int i = 0; i < aiMessage.toolExecutionRequests().size(); i++) {
+                // Add tool call attributes here if needed
+            }
+        }
+    }
+    
+    private String mapMessageRole(ChatMessageType type) {
+        switch (type) {
+            case SYSTEM:
+                return "system";
+            case USER:
+                return "user";
+            case AI:
+                return "assistant";
+            case TOOL_EXECUTION_RESULT:
+                return "tool";
+            default:
+                return type.toString().toLowerCase();
+        }
+    }
+    
+    private List<Map<String, Object>> convertMessages(List<ChatMessage> messages) {
+        List<Map<String, Object>> result = new ArrayList<>();
+        
+        for (ChatMessage message : messages) {
+            Map<String, Object> messageMap = new HashMap<>();
+            
+            switch (message.type()) {
+                case SYSTEM:
+                    messageMap.put("role", "system");
+                    messageMap.put("content", message.text());
+                    break;
+                case USER:
+                    messageMap.put("role", "user");
+                    messageMap.put("content", message.text());
+                    break;
+                case AI:
+                    messageMap.put("role", "assistant");
+                    messageMap.put("content", message.text());
+                    if (message instanceof AiMessage) {
+                        AiMessage aiMessage = (AiMessage) message;
+                        if (aiMessage.toolExecutionRequests() != null && !aiMessage.toolExecutionRequests().isEmpty()) {
+                            messageMap.put("tool_calls", aiMessage.toolExecutionRequests());
+                        }
+                    }
+                    break;
+                case TOOL_EXECUTION_RESULT:
+                    messageMap.put("role", "tool");
+                    messageMap.put("content", message.text());
+                    break;
+            }
+            
+            result.add(messageMap);
+        }
+        
+        return result;
+    }
+    
+    private String detectProvider(String modelName) {
+        if (modelName == null) {
+            return null;
+        }
+        
+        String lowerModel = modelName.toLowerCase();
+        if (lowerModel.contains("gpt") || lowerModel.contains("dall-e") || 
+            lowerModel.contains("whisper") || lowerModel.contains("text-embedding") ||
+            lowerModel.contains("davinci") || lowerModel.contains("curie") || 
+            lowerModel.contains("babbage") || lowerModel.contains("ada")) {
+            return "openai";
+        } else if (lowerModel.contains("claude")) {
+            return "anthropic";
+        } else if (lowerModel.contains("gemini") || lowerModel.contains("palm")) {
+            return "google";
+        } else if (lowerModel.contains("llama") || lowerModel.contains("codellama")) {
```

Comment:
this seems brittle since these can be deployed on other providers. Also we should probably use enums just to be save